### PR TITLE
Sac tweaks

### PR DESCRIPTION
**Pull request by dosssman**

### CODE REVIEW

File: cleanrl/sac_continuous_action.py [hunk 1]
```diff
@@ -104,26 +127,56 @@
     env = Monitor(env, f'videos/{experiment_name}')
 
 # ALGO LOGIC: initialize agent here:
+# Weight init scheme
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:130]: The code seems a little messy for the initialization method. Consider taking a look at how this is done in PPO https://github.com/vwxyzjn/cleanrl/blob/65d0e4a1c3be60fd590b30108869b764990a2cb3/cleanrl/ppo_continuous_action.py#L221.

File: cleanrl/sac_continuous_action.py [hunk 2]
```diff
@@ -148,12 +201,16 @@ def to(self, device):
...
-        self.fc2 = nn.Linear(120, 84)
-        self.fc3 = nn.Linear(84, 1)
+        self.fc1 = nn.Linear(input_shape+output_shape, 256)
+        self.fc2 = nn.Linear(256, 128)
+        self.fc3 = nn.Linear(128, 1)
+        self.apply(weights_and_bias_init_)
 
     def forward(self, x, a):
-        x = preprocess_obs_fn(x)
+        if not isinstance(x, torch.Tensor):
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:210]: Consider looking at how it is done in https://github.com/vwxyzjn/cleanrl/blob/65d0e4a1c3be60fd590b30108869b764990a2cb3/cleanrl/td3_continuous_action.py#L143

File: cleanrl/sac_continuous_action.py [hunk 3]
```diff
@@ -48,22 +51,42 @@
                         help="the wandb's project name")
     parser.add_argument('--wandb-entity', type=str, default=None,
                         help="the entity (team) of wandb's project")
-    
+    parser.add_argument('--autotune', type=lambda x:bool(strtobool(x)), default=False, nargs='?', const=True,
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:54]: Even if it doesn’t make much difference, consider turning on the auto tune by default

File: cleanrl/sac_continuous_action.py [hunk 4]
```diff
@@ -192,88 +249,116 @@ def sample(self, n):
...
+    # ALGO LOGIC: put action logic here
+    if global_step < args.learning_starts:
+        action = env.action_space.sample()
+    else:
+        action, _, _ = pg.get_action([obs])
+        action = action.tolist()[0]
+
+    # TRY NOT TO MODIFY: execute the game and log data.
+    next_obs, rew, done, _ = env.step(action)
+    rb.put((obs,action,rew,next_obs,done))
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:281]: There should be a space between the commas

File: cleanrl/sac_continuous_action.py [hunk 5]
```diff
@@ -192,88 +249,116 @@ def sample(self, n):
...
+
+    # ALGO LOGIC: put action logic here
+    if global_step < args.learning_starts:
+        action = env.action_space.sample()
+    else:
+        action, _, _ = pg.get_action([obs])
+        action = action.tolist()[0]
+
+    # TRY NOT TO MODIFY: execute the game and log data.
+    next_obs, rew, done, _ = env.step(action)
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:280]: rew should be reward

File: cleanrl/sac_continuous_action.py [hunk 6]
```diff
@@ -192,88 +249,116 @@ def sample(self, n):
...
-    writer.add_scalar("losses/soft_q_value_2_loss", qf2_losses[:step+1].mean(), global_step)
-    writer.add_scalar("losses/policy_loss", policy_losses[:step+1].mean(), global_step)
+
+    if done:
+        global_episode += 1 # Outside the loop already means the epsiode is done
+
+        writer.add_scalar("charts/episode_reward", episode_reward, global_step)
+        writer.add_scalar("charts/episode_length", episode_length, global_step)
+
+        if len(rb.buffer) > args.batch_size: # Makes sure loss and such are actually available.
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:344]: Consider moving the metrics recording scripts out of the`if done:` block, Additionally only record the loss every 100 times step. See example https://github.com/vwxyzjn/cleanrl/blob/65d0e4a1c3be60fd590b30108869b764990a2cb3/cleanrl/td3_continuous_action.py#L244

File: cleanrl/sac_continuous_action.py [hunk 7]
```diff
@@ -8,14 +8,17 @@
...
 
-from cleanrl.common import preprocess_obs_space, preprocess_ac_space
 import argparse
 from distutils.util import strtobool
 import collections
 import numpy as np
 import gym
 from gym.wrappers import TimeLimit, Monitor
 import pybullet_envs
+try:
```
[the remark this answers is not in the document]
Comment (dosssman) [reply cleanrl/sac_continuous_action.py:18]: Beside out of the box support for Mujoco envs, none.
I guess it can be added on a per-need basis later. Removing it.

File: cleanrl/sac_continuous_action.py (hunk 1 above)
Comment (dosssman) [reply cleanrl/sac_continuous_action.py]: Not just a little bit messy haha. Tidied up a little bit.
Removing the `if isinstance(...)` line would require to change to model definition into something like `self.net = nn.Sequential(...)` in both actor and critic ...

File: cleanrl/sac_continuous_action.py [hunk 8]
```diff
@@ -89,8 +108,8 @@
 env.seed(args.seed)
 env.action_space.seed(args.seed)
 env.observation_space.seed(args.seed)
-input_shape, preprocess_obs_fn = preprocess_obs_space(env.observation_space, device)
-output_shape = preprocess_ac_space(env.action_space)
+input_shape = env.observation_space.shape[0]
+output_shape = env.action_space.shape[0]
 # respect the default timelimit
 assert isinstance(env.action_space, Box), "only continuous action space is supported"
 assert isinstance(env, TimeLimit) or int(args.episode_length), "the gym env does not have a built in TimeLimit, please specify by using --episode-length"
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:115]: This should not be needed anymore

File: cleanrl/sac_continuous_action.py [hunk 9]
```diff
@@ -104,26 +123,43 @@
     env = Monitor(env, f'videos/{experiment_name}')
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:124]: The new line here is not Necessary

File: cleanrl/sac_continuous_action.py [hunk 10]
```diff
@@ -104,26 +123,43 @@
...
+LOG_STD_MAX = 2
+LOG_STD_MIN = -5
+
+def layer_init(layer, weight_gain=1, bias_const=0):
+    if isinstance(layer, nn.Linear):
+        if args.weights_init == "xavier":
+            torch.nn.init.xavier_uniform_(layer.weight, gain=weight_gain)
+        elif args.weights_init == "orthogonal":
+            torch.nn.init.orthogonal_(layer.weight, gain=weight_gain)
+
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:135]: The new line Here is not needed

File: cleanrl/sac_continuous_action.py [hunk 11]
```diff
@@ -192,88 +229,116 @@ def sample(self, n):
...
+            writer.add_scalar("losses/qf_loss", qf_loss.item(), global_step)
+            writer.add_scalar("losses/policy_loss", policy_loss.item(), global_step)
+            writer.add_scalar("losses/alpha", alpha, global_step)
+            if args.autotune:
+                writer.add_scalar("losses/alpha_loss", alpha_loss.item(), global_step)
+
+        # Reseting what need to be
+        obs, done = env.reset(), False
+        episode_reward, episode_length = 0., 0
+
```
Comment (vwxyzjn) [inline cleanrl/sac_continuous_action.py:340]: Some new lines over here are not necessary

File: cleanrl/sac_continuous_action.py (hunk 8 above)
Comment (dosssman) [reply cleanrl/sac_continuous_action.py:115]: The time limit ?

Comment (vwxyzjn) [reply cleanrl/sac_continuous_action.py:115]: Yes
